pkg/controller/v1beta1/inferenceservice/reconcilers/deployment/deployment_reconciler.go

```diff
@@ -204,50 +312,102 @@ func setDefaultDeploymentSpec(spec *appsv1.DeploymentSpec) {
+				})
+				deploymentSpec.Template.Spec.Containers[i].Env = container.Env
+			}
+			log.Info("Added env variable to container",
+				"envName", envName,
+				"envValue", envValue,
+				"containerName", containerName)
+			return
+		}
 	}
-	log.Info("deployment reconcile", "checkResult", checkResult, "err", err)
+	log.Info("Container not found in DeploymentSpec", "containerName", containerName)
+}
 
-	var opErr error
-	switch checkResult {
-	case constants.CheckResultCreate:
-		opErr = r.client.Create(context.TODO(), r.Deployment)
-	case constants.CheckResultUpdate:
-		curJson, err := json.Marshal(deployment)
-		if err != nil {
-			return nil, err
-		}
+func addGPUResourceToDeployment(deployment *appsv1.Deployment, targetContainerName string, tensorParallelSize string) {
+	for i, container := range deployment.Spec.Template.Spec.Containers {
+		if container.Name == targetContainerName {
+			// Initialize Limits map if it's nil
+			if container.Resources.Limits == nil {
+				deployment.Spec.Template.Spec.Containers[i].Resources.Limits = make(map[corev1.ResourceName]resource.Quantity)
+			}
 
-		// To avoid the conflict between HPA and Deployment,
-		// we need to remove the Replicas field from the deployment spec
-		modDeployment := r.Deployment.DeepCopy()
-		modDeployment.Spec.Replicas = nil
+			// Assign the tensorParallelSize value to the GPU "nvidia.com/gpu" resource limits
+			deployment.Spec.Template.Spec.Containers[i].Resources.Limits[constants.NvidiaGPUResourceType] = resource.MustParse(tensorParallelSize)
 
-		modJson, err := json.Marshal(modDeployment)
-		if err != nil {
-			return nil, err
+			// Initialize Requests map if it's nil
+			if container.Resources.Requests == nil {
+				deployment.Spec.Template.Spec.Containers[i].Resources.Requests = make(map[corev1.ResourceName]resource.Quantity)
+			}
+
+			// Assign the tensorParallelSize value to the GPU "nvidia.com/gpu" resource requests
+			deployment.Spec.Template.Spec.Containers[i].Resources.Requests[constants.NvidiaGPUResourceType] = resource.MustParse(tensorParallelSize)
+			break
 		}
-		// Generate the strategic merge patch between the current and modified JSON
-		patchByte, err := strategicpatch.StrategicMergePatch(curJson, modJson, appsv1.Deployment{})
+	}
+}
+
+// Reconcile ...
+func (r *DeploymentReconciler) Reconcile() ([]*appsv1.Deployment, error) {
+	for _, deployment := range r.DeploymentList {
+		// Reconcile Deployment
+		checkResult, _, err := r.checkDeploymentExist(r.client, deployment)
 		if err != nil {
 			return nil, err
 		}
+		log.Info("deployment reconcile", "checkResult", checkResult, "err", err)
 
-		// Patch the deployment object with the strategic merge patch
-		opErr = r.client.Patch(context.TODO(), deployment, client.RawPatch(types.StrategicMergePatchType, patchByte))
+		var opErr error
+		switch checkResult {
+		case constants.CheckResultCreate:
+			opErr = r.client.Create(context.TODO(), deployment)
+		case constants.CheckResultUpdate:
```

Comment:
Currently, the implementation uses `Deployment`, so it is very similar to the existing approach. Since upgrades are handled using `Deployment's rollingUpdate`, I believe there will be no downtime, as the existing pods will continue handling requests until the new pods are ready.

Regarding the shared PVC, as explained in another comment, we support the PVC protocol, meaning that the management of models within the PVC is the user's responsibility. Therefore, I don't think a dedicated reconciler is necessary.

Reply:
I haven't tested the upgrade process, so I didn’t realize there could be this issue. For the upgrade to work correctly, the worker node needs to accurately know the IP of its head pod. What would be the best approach for this?

One idea I have is to pass the object_id of the isvc to each deployment. This way, both the head and worker nodes would be paired together. What do you think?

Reply:
I changed the logic a little bit to use inferenceSerivce generation. To upgrade, it needs to update isvc. Then the generation will be changed. 
Using this behavior, I add this information into deployment and service. So head svc name will be `XXX-HEAD-$GENERATION` and it will select the label ``isvc.generation`: XX` and `app=isvc.$isvc_name-predictor,`
Then from worker node will have `HEAD_SVC` to pointing to the head svc.

Plus, I added cleanup methods for head svc if there are more then 3svcs created for each isvc generation.